### PR TITLE
AK: Add IPv6Address class

### PR DESCRIPTION
**Pull request by tomuta**
This is the IPv6 counter part to the IPv4Address class and implements
parsing strings into a in6_addr and formatting one as a string. It
supports the address compression scheme as well as IPv4 mapped
addresses.

### CODE REVIEW

File: AK/IPv6Address.h [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        m_data[11] = 0xff;
+        m_data[12] = ipv4_address[0];
+        m_data[13] = ipv4_address[1];
+        m_data[14] = ipv4_address[2];
+        m_data[15] = ipv4_address[3];
+    }
+
+    constexpr u16 operator[](int i) const { return group(i); }
+
+    String to_string() const
```
Comment (IdanHo) [inline AK/IPv6Address.h:51]: This will not compile if it's ever included in the Kernel.

File: Userland/Libraries/LibC/arpa/inet.cpp [hunk 2]
```diff
@@ -13,49 +14,68 @@ extern "C" {
...
-    unsigned b;
-    unsigned c;
-    unsigned d;
-    int count = sscanf(src, "%u.%u.%u.%u", &a, &b, &c, &d);
-    if (count != 4) {
-        errno = EINVAL;
-        return 0;
+    if (af == AF_INET) {
+        unsigned a, b, c, d;
+        int count = sscanf(src, "%u.%u.%u.%u", &a, &b, &c, &d);
```
Comment (IdanHo) [inline Userland/Libraries/LibC/arpa/inet.cpp:46]: This looks good, but how about adding another commit replacing the ipv4 part to use AK::IPv4Address for consistency?

Comment (tomuta) [reply Userland/Libraries/LibC/arpa/inet.cpp:46]: Yeah sounds like something we should eventually do. Maybe I'll do that in a separate PR some other day.

File: AK/IPv6Address.h [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        return builder.string_view();
+#endif
+    }
+
+    static Optional<IPv6Address> from_string(StringView string)
+    {
+        if (string.is_null())
+            return {};
+
+        const auto parts = string.split_view(':', true);
```
Comment (linusg) [inline AK/IPv6Address.h:128]: East const pls

File: AK/IPv6Address.h [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    {
+        if (is_ipv4_mapped())
+            return IPv4Address(m_data[12], m_data[13], m_data[14], m_data[15]);
+        return {};
+    }
+
+private:
+    constexpr u16 group(int i) const
+    {
+        VERIFY(i >= 0 && i < 8);
```
Comment (linusg) [inline AK/IPv6Address.h:257]: Can we not use an unsigned data type instead of VERIFY'ing? `size_t`?

File: AK/IPv6Address.h [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    in6_addr_t m_data {};
+};
+
+static_assert(sizeof(IPv6Address) == 16);
+
+template<>
+struct Traits<IPv6Address> : public GenericTraits<IPv6Address> {
+    static constexpr unsigned hash(const IPv6Address& address)
```
Comment (linusg) [inline AK/IPv6Address.h:268]: East const

File: AK/IPv6Address.h [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+                h = pair_int_hash(h, two_groups);
+        }
+        return h;
+    }
+};
+
+#ifdef KERNEL
+template<>
+struct Formatter<IPv6Address> : Formatter<ErrorOr<NonnullOwnPtr<Kernel::KString>>> {
+    ErrorOr<void> format(FormatBuilder& builder, IPv6Address value)
```
Comment (linusg) [inline AK/IPv6Address.h:285]: I don't think we need to copy `value` here

File: AK/IPv6Address.h [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    {
+        return Formatter<ErrorOr<NonnullOwnPtr<Kernel::KString>>>::format(builder, value.to_string());
+    }
+};
+#else
+template<>
+struct Formatter<IPv6Address> : Formatter<String> {
+    ErrorOr<void> format(FormatBuilder& builder, IPv6Address value)
```
Comment (linusg) [inline AK/IPv6Address.h:293]: Likewise

File: Tests/AK/TestIPv6Address.cpp [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    EXPECT(mapped_address.is_ipv4_mapped());
+    EXPECT_EQ(ipv4_address_to_map, mapped_address.ipv4_mapped_address().value());
+    EXPECT_EQ("::ffff:192.168.0.1"sv, mapped_address.to_string());
+    EXPECT_EQ(IPv4Address(192, 168, 1, 9), IPv6Address::from_string("::FFFF:192.168.1.9"sv).value().ipv4_mapped_address().value());
+    EXPECT(!IPv6Address::from_string("::abcd:192.168.1.9"sv).has_value());
+}
+
+TEST_CASE(should_make_empty_optional_from_bad_string)
+{
+    const auto addr = IPv6Address::from_string("bad string"sv);
```
Comment (linusg) [inline Tests/AK/TestIPv6Address.cpp:96]: East const

File: Tests/AK/TestIPv6Address.cpp [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+    EXPECT(!addr.has_value());
+}
+
+TEST_CASE(should_make_empty_optional_from_out_of_range_values)
+{
+    const auto addr = IPv6Address::from_string("::10000"sv);
```
Comment (linusg) [inline Tests/AK/TestIPv6Address.cpp:103]: East const

File: AK/IPv6Address.h [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+            group++;
+        }
+
+        return IPv6Address(addr);
+    }
+
+    constexpr in6_addr_t const& to_in6_addr_t() const { return m_data; }
+
+    constexpr bool operator==(const IPv6Address& other) const = default;
+    constexpr bool operator!=(const IPv6Address& other) const = default;
```
Comment (linusg) [inline AK/IPv6Address.h:227]: East const

File: Tests/AK/TestIPv6Address.cpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    static_assert(addr.is_zero());
+
+    EXPECT(addr.is_zero());
+}
+
+TEST_CASE(should_construct_from_c_array)
+{
+    constexpr auto addr = [] {
+        const u8 a[16] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 };
```
Comment (linusg) [inline Tests/AK/TestIPv6Address.cpp:24]: East const

**Review by linusg — COMMENTED** [review]
Mostly LGTM, some questions. Please remember to use east const :)

**Pull request merged**